### PR TITLE
Amazon push

### PR DESCRIPTION
**Pull request by JohnOKane**

### CODE REVIEW

File: native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmPushSupport.java [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+                String registrationId = adm.getRegistrationId();
+                if (SwrveAdmHelper.isNullOrEmpty(registrationId)) {
+                    Log.i(TAG, "adm.getRegistrationId() returned null. Will call adm.startRegister().");
+                    adm.startRegister();
+                } else {
+                    Log.i(TAG, "adm.getRegistrationId() returned: " + registrationId);
+                    notifySDKOfRegistrationId(gameObject, registrationId);
+                }
+                sdkIsReadyToReceivePushNotifications(activity);
+            } catch (Throwable exp) {
```
Comment (Sergio-Mira) [inline native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmPushSupport.java]: Change to Exception if changed on the native SDK

**Comment by Sergio-Mira** [comment]
🚀 !

**Comment by Sergio-Mira** [comment]
🚀 !

**Comment by JohnOKane** [comment]
@seaders Attempt to involve as participant.

File: native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmIntentService.java [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+                    pushTitle = appTitle.toString();
+                }
+            }
+            if (SwrveAdmHelper.isNullOrEmpty(pushTitle)) {
+                pushTitle = "Configure your app title";
+            }
+        }
+
+        // Build notification
+        NotificationCompat.Builder mBuilder = new NotificationCompat.Builder(this)
```
Comment (seaders) [inline native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmIntentService.java]: ~(pedantic) "mBuilder".  I'm not keen on the "m" prefix at the best of times, but really don't know why it's here at all.

File: native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmPushSupport.java [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+            UnityPlayer.UnitySendMessage(gameObject, "OnNotificationReceivedADM", serializedNotification.toString());
+        }
+    }
+
+    private static void notifySDKOfOpenedNotification(Context context, SwrveNotification notification) {
+        String gameObject = getGameObject(context);
+        String serializedNotification = notification.toJson();
+        if (serializedNotification != null) {
+            UnityPlayer.UnitySendMessage(gameObject, "OnOpenedFromPushNotificationADM", serializedNotification.toString());
+        }
```
Comment (seaders) [inline native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmPushSupport.java:178]: Externalise these strings to the top of the class?  With reference to where they're calling on the Unity side, which specific class?

File: unity3d/Assets/Swrve/SwrveSDK/SwrveNativeAndroid.cs [hunk 4]
```diff
@@ -233,6 +244,63 @@ private void RequestGooglePlayAdvertisingId(MonoBehaviour container)
...
+
+		        androidADMPlugin = new AndroidJavaClass(pluginPackageName);
+		        if (androidADMPlugin == null) {
+		            SwrveLog.LogError("Found class, but unable to construct AndroidJavaClass: " + jniPluginClassName);
+		            return;
+		        }
+
+		        // Check that the version is the same
+		        int testPluginVersion = androidADMPlugin.CallStatic<int>("getVersion");
+
```
Comment (seaders) [inline unity3d/Assets/Swrve/SwrveSDK/SwrveNativeAndroid.cs]: Externalise strings. Variable name "testPluginVersion"?

File: unity3d/Assets/Swrve/SwrveSDK/SwrveNativeAndroid.cs [hunk 5]
```diff
@@ -347,6 +437,33 @@ public void NotificationReceived(string notificationJson)
...
+    /// </param>
+    public void NotificationReceivedADM(string notificationJson)
+    {
+        Dictionary<string, object> notification = (Dictionary<string, object>)Json.Deserialize (notificationJson);
+        if (androidADMPlugin != null && notification != null) {
+            string pushId = GetPushId(notification);
+            if (pushId != null) {
+                // Acknowledge the received notification
+                androidADMPlugin.CallStatic("sdkAcknowledgeReceivedNotification", pushId);
+            }
```
Comment (seaders) [inline unity3d/Assets/Swrve/SwrveSDK/SwrveNativeAndroid.cs:459]: Externalise string.

File: unity3d/Assets/Swrve/SwrveSDK/SwrveNativeAndroid.cs [hunk 6]
```diff
@@ -397,6 +514,40 @@ public void OpenedFromPushNotification(string notificationJson)
...
+    /// </param>
+    public void OpenedFromPushNotificationADM(string notificationJson)
+    {
+        Dictionary<string, object> notification = (Dictionary<string, object>)Json.Deserialize (notificationJson);
+        string pushId = GetPushId(notification);
+        SendPushEngagedEvent(pushId);
+        if (pushId != null && androidADMPlugin != null) {
+            // Acknowledge the received notification
+            androidADMPlugin.CallStatic("sdkAcknowledgeOpenedNotification", pushId);
+        }
```
Comment (seaders) [inline unity3d/Assets/Swrve/SwrveSDK/SwrveNativeAndroid.cs:537]: Externalise string.

File: native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmIntentService.java [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+            //Try get de-dupe cache size
+            int pushIdCacheSize = msg.getInt(SwrveAdmHelper.PUSH_ID_CACHE_SIZE_KEY, DEFAULT_PUSH_ID_CACHE_SIZE);
+
+            //No duplicate found. Update the cache.
+            updateRecentNotificationIdCache(recentIds, curId, pushIdCacheSize);
+
+            final SharedPreferences prefs = SwrveAdmPushSupport.getAdmPreferences(getApplicationContext());
+            String activityClassName = prefs.getString(SwrveAdmPushSupport.PROPERTY_ACTIVITY_NAME, null);
+            if (SwrveAdmHelper.isNullOrEmpty(activityClassName)) {
+                activityClassName = "com.unity3d.player.UnityPlayerNativeActivity";
```
Comment (seaders) [inline native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmIntentService.java]: Externalise string.

File: native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmIntentService.java (hunk 2 above)
Comment (JohnOKane) [reply native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmIntentService.java]: Basically as found here: https://github.com/Swrve/swrve-unity-sdk/blob/release-4_7/native/android/SwrveSDKPushSupport/src/main/java/com/swrve/unity/gcm/SwrveGcmIntentService.java#L171 and elsewhere in that file. Code taking all conventions and much of the implementation from there.

**Comment by seaders** [comment]
Other than a few small bits, looks good to me.  Before merging in, I'd grab a few binaries built from CI to make 100% sure all push things that were working still are.

File: native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmIntentService.java (hunk 2 above)
Comment (seaders) [reply native/android/SwrveSDKPushSupport/src/amazon/java/com/swrve/unity/adm/SwrveAdmIntentService.java]: Yeah... I wouldn't agree with it there, either!  "m" is old school C / C++ standard for member variable, and I didn't even like them back then.  This ain't a member variable, and this is Java.

**Pull request merged**